sound/soc/sof/topology.c

```diff
@@ -2523,23 +2540,23 @@ static int sof_widget_unload(struct snd_soc_component *scomp,
 			kfree(dai->dai_config);
 			list_del(&dai->list);
 		}
+
+		pipeline = snd_sof_pipeline_find(sdev, swidget->pipeline_id);
+
 		break;
 	case snd_soc_dapm_scheduler:
-
-		/* power down the pipeline schedule core */
 		pipeline = swidget->private;
-		ret = snd_sof_dsp_core_power_down(sdev, 1 << pipeline->core);
-		if (ret < 0)
-			dev_err(scomp->dev, "error: powering down pipeline schedule core %d\n",
-				pipeline->core);
-
-		/* update enabled cores mask */
-		sdev->enabled_cores_mask &= ~(1 << pipeline->core);
 
 		break;
 	default:
+		pipeline = snd_sof_pipeline_find(sdev, swidget->pipeline_id);
+
```

Comment:
@lyakh Is this safe? I mean there are lots of kinds of widgets and with this patch we go from powering off from dapm_schduler to powering off for any swidget unload? Can this be right?

Reply:
@kv2019i so far I'm unaware of any cases where we would only unload some widgets and keep others. If we already have such cases, then this isn't right. The next possibility where we would do that, that I am aware of is VirtIO SOF. Then when a guest is shut down (or suspended), its topology gets unloaded, while the rest is kept. For that to work we'd need per-core refcounts. We can decide whether to do that now or to first take this fix and implement refcounting later when we have any real life dynamic topology cases.

Reply:
I had the same reaction than Kai. This would turn off the core for the first widget that gets unloaded, but shouldn't we power down for the the last widget to be on the safe side?

Reply:
@plbossart Sure, I'd be glad to do that, but for that we need refcounting. I'll double-check on Monday, and maybe I'll find a location after the topology is unloaded to just power down all the cores that are enabled in the mask.